chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! The ChaCha20 stream cipher from RFC8439.
+
+/// The first four words (32-bit) of the ChaCha stream cipher state are constants.
+const WORD_1: u32 = 0x61707865;
+const WORD_2: u32 = 0x3320646e;
+const WORD_3: u32 = 0x79622d32;
+const WORD_4: u32 = 0x6b206574;
+
+/// The quarter round indexes of a ChaCha block function.
+const CHACHA_ROUND_INDICIES: [(usize, usize, usize, usize); 8] = [
+    // The first 4 rounds are columns of the 4x4 matrix.
+    (0, 4, 8, 12),
+    (1, 5, 9, 13),
+    (2, 6, 10, 14),
+    (3, 7, 11, 15),
+    // The next 4 rounds are diagonals of the 4x4 matrix.
+    (0, 5, 10, 15),
+    (1, 6, 11, 12),
+    (2, 7, 8, 13),
+    (3, 4, 9, 14),
+];
+
+/// The cipher's block size is 64 bytes.
+const CHACHA_BLOCKSIZE: usize = 64;
+
+/// A 256-bit secret session key shared by the parties communicating.
+#[derive(Clone, Copy, Debug)]
+pub struct SessionKey([u8; 32]);
+
+impl SessionKey {
+    /// Create a new session key.
+    pub const fn new(key: [u8; 32]) -> Self {
+        SessionKey(key)
+    }
+}
+
+/// A 96-bit initialization vector (IV), or nonce.
+#[derive(Clone, Copy, Debug)]
+pub struct Nonce([u8; 12]);
+
+impl Nonce {
+    /// Create a new nonce.
+    pub const fn new(nonce: [u8; 12]) -> Self {
+        Nonce(nonce)
+    }
+}
+
+/// The 512-bit cipher state is chunk'd up into 16 32-bit words.
+///
+/// The 16 words can be visualized as a 4x4 matrix:
+///
+///   0   1   2   3
+///   4   5   6   7
+///   8   9  10  11
+///   12  13  14  15
+#[derive(Clone, Copy, Debug)]
+struct State([u32; 16]);
+
+impl State {
+    /// New prepared state.
+    const fn new(key: SessionKey, nonce: Nonce, count: u32) -> Self {
+        let mut state: [u32; 16] =
+            [WORD_1, WORD_2, WORD_3, WORD_4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+
+        // The next two rows (8 words) are based on the session key.
+        // Using a while loop here to keep the function constant.
```

Comment:
const, not constant. "Constant" is undefined and potentially suggests constant-time which it happens to be but it's entirely irrelevant here.

Reply:
Makes sense, thanks